### DISCUSSION TITLE
Ansible community.clickhouse collection with modules is published

### DISCUSSION THREAD

**Discussion by Andersson007** [ClickHouse/ClickHouse#59222] [Show and tell]
Hello, ClichHouse community!

I didn't find any modules for ClickHouse, so created the ~~MVP~~ ``community.clickhouse`` Ansible collection.
Version 0.1.0 (see the comment for release announcements) is published in on [Galaxy](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/).

It contains a few modules, see the [Galaxy doc page](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/docs/) for details.

Examples:
```
- name: Get server information
  register: result
  community.clickhouse.clickhouse_info:

- name: Print server information
  ansible.builtin.debug:
    var: result.result
```
The play will print (the output made short):
```
}
    "driver": {
        "version": "0.2.6"
    },
    "version": {
        "build": 59,
        "feature": 12,
        "maintenance": 2,
        "raw": "23.12.2.59",
        "type": null,
        "year": 23
    }
    "roles": {
        "accountant": {
            "id": "4d4b281d-2a20-2bee-475b-ff9b2290d98e",
            "storage": "local_directory"
        }
    },
    "settings": {
        "add_http_cors_header": {
            "changed": 0,
            "default": "0",
            "description": "Write add http CORS header.",
            "is_obsolete": 0,
            "max": null,
            "min": null,
            "readonly": 0,
            "value": "0"
        },
        ...
    }
    "databases": {
        "INFORMATION_SCHEMA": {
            "comment": "",
            "data_path": "/var/lib/clickhouse/",
            "engine": "Memory",
            "uuid": "00000000-0000-0000-0000-000000000000"
        },
        ...
    }
    "users": {
        "default": {
            "auth_params": "{}",
            "auth_type": "plaintext_password",
            "default_roles_all": 1,
            "default_roles_except": [],
            "default_roles_list": [],
            "host_ip": [
                "::/0"
            ],
            "host_names": [],
            "host_names_like": [],
            "host_names_regexp": [],
            "id": "94309d50-4f52-5250-31bd-74fecac179db",
            "storage": "users_xml"
        }
    },
    "clusters": {
        ...
    },
}

```
With _client module, you can run any queries, for instance:
```
# Note: it runs SELECT version() just for the sake of example.
# You can get it with the task above much more conveniently.
- name: Query DB using non-default user & DB to connect to
  register: result
  community.clickhouse.clickhouse_client:
    execute: SELECT version()
    login_host: localhost
    login_user: alice
    login_db: foo
    login_password: my_password

- name: Print returned server version
  ansible.builtin.debug:
    var: result.result
```

Provided that there's an interest from the community and feedback,
more functionality and modules can be added in future (like modules for creating users
or databases, setting up the server configuration, etc. idempotently & more conveniently than using the `clickhouse_client` module).

Users and contributors are welcome: I'd be happy to help. No feedback, no continuation:)
Thanks:)

**Comment by Andersson007** [comment]
FYI @aleksvagachev fixed the `clickhouse_info` module to support earlier versions (21.3.4.25). So if there were some failures because of the fields missed in earlier version, now it should work at least for 21.3.4.25.
I published version 0.1.1 on Galaxy. You can quickly upgrade by running `ansible-galaxy collection install community.clickhouse --upgrade`

**Comment by aleksvagachev** [comment]
Version 0.3.0 has been released for the community.clickhouse collection.
- The collection has been tested for versions 21.8.15.7, 22.8.9.24, 23.8.9.54
- Added a module for creating/deleting databases
- Added the ability to collect more system information

The details are on the [Galaxy page](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/).

I published version 0.3.0 on Galaxy. You can quickly upgrade by running `ansible-galaxy collection install community.clickhouse --upgrade`

**Comment by aleksvagachev** [comment]
Version 0.4.0 has been released for the community.clickhouse collection.

- Added new functionality for creating/deleting/renaming databases
- Added a module for creating/deleting users
- Added the ability to collect more system information

The details are on the [Galaxy page](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/).

Version 0.4.0 has been published on Galaxy. You can quickly upgrade by running ansible-galaxy collection install community.clickhouse --upgrade

The collection exceeded 1200 downloads from Galaxy but we have almost no starts in the repo.
If you like the project and want to motivate us, please put a start on the [repo](https://github.com/ansible-collections/community.clickhouse)!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.5.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v050) has just been released!
If you're a collection user and like the initiative, please add a star in our [repo](https://github.com/ansible-collections/community.clickhouse) to motivate us a bit:)

You can quickly upgrade by running `ansible-galaxy collection install community.clickhouse --upgrade`

The release:
- adds an important option `set_settings` for the clickhouse_client module that now allows you to set arbitrary session settings before running the main statement, for example to disable `flatten_nested` to be able to run statements against nested datastructures add to your task:

```
    set_settings:
      flatten_nested: 0
```
- adds improvements to the clickhouse_user module
- adds the clickhouse_role MVP version: its very raw but we included it to give intested people an opportunity to improve it:)

Any contributions and feedback are very welcome! See the [open issues](https://github.com/ansible-collections/community.clickhouse/issues) and our [contributor guide](https://github.com/ansible-collections/community.clickhouse/blob/main/CONTRIBUTING.md#code-contributions) for details.

Thanks for your attention!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.6.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v060) has been released! Please [give it a star](https://github.com/ansible-collections/community.clickhouse/tree/main) to support the initiative.

It includes the following minor changes:

- clickhouse_info - add the ``roles`` field to user information.
- clickhouse_user - add the ``default_roles_mode`` argument to specify how to handle roles passed through ``default_roles`` argument.
- clickhouse_user - add the ``default_roles`` argument to set default roles.
- clickhouse_user - add the ``roles_mode`` argument to specify how to handle roles passed through ``roles`` argument.
- clickhouse_user - add the ``roles`` argument to grant roles.

So from now on you can:
1. Fetch user roles info using `clickhouse_info` module
2. Grant roles to a user and set default roles, for example

```
- name: Append roles and set a default role for a user
  clickhouse.clickhouse_user:
    name: bob
    roles:
    - accountant
    - sales
    roles_mode: append
    default_roles:
    - accountant
 ```
See the [Documentation](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/docs/) tab on the galaxy page for more examples.

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection:
- Version [0.7.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v070) has been released! Among a few minor changes, it ships a new info module to fetch information from YAML and XML config files and return it as JSON.
- We'd also like to remind about the [ClickHouse Ansible collection team](https://forum.ansible.com/g/ClickHouseTeam) on the Ansible forum that welcome folks to join.
- For the ClickHouse automation related questions, please use the `clickhouse` tag on the [forum](https://forum.ansible.com/).
- We are planning to submit the collection for inclusion in the Ansible community package soon!
- Your starts matter to us! Please star the [repo](https://github.com/ansible-collections/community.clickhouse/tree/main) if you like the initiative.

**Reply by wsargent** [reply]
Thank you for doing this.  I'm just setting up a hobby project and your recipe is very useful. :-)

**Reply by Andersson007** [reply]
@wsargent thanks for the feedback! Any issues reported in the [repo](https://github.com/ansible-collections/community.clickhouse), would be very appreciated too!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse) collection version [0.8.0](https://github.com/ansible-collections/community.clickhouse/releases) has been released with:
- dropping support for ClickHouse versions 21 and 22
- improvements to the clickhouse_info module
- a bugfix for the clickhouse_user module

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse) collection version [0.8.1](https://github.com/ansible-collections/community.clickhouse/releases/tag/0.8.1) has been released!
- It ensures the collection works with the upcoming ansible-core 2.19 version. Make sure you've upgraded the collection before upgrading ansible-core to 2.19 after its release.

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse) collection version [0.8.2](https://github.com/ansible-collections/community.clickhouse/releases/tag/0.8.2) has been released!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse) collection version [0.8.3](https://github.com/ansible-collections/community.clickhouse/releases/tag/0.8.3) has been released!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse) collection version [0.8.4](https://github.com/ansible-collections/community.clickhouse/releases/tag/0.8.4) has been released!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse) collection version [0.8.5](https://github.com/ansible-collections/community.clickhouse/releases/tag/0.8.5) has been released!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.9.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v090) has been released!

It contains a new module: `clickhouse_grants` - Manage grants for ClickHouse users and roles

So, I think we are close to submitting the collection for inclusion in the Ansible community package (that one you install via `pip install ansible` or `dnf/apt install ansible`). The collection now contains the following modules:
- [clickhouse_cfg_info](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_cfg_info/) module:	Retrieves ClickHouse config file content and returns it as JSON
- [clickhouse_client](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_client/) module:	Execute queries in a ClickHouse database using the clickhouse-driver Client interface
- [clickhouse_db](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_db/)	module:	Creates or removes a ClickHouse database using the clickhouse-driver Client interface
- [clickhouse_grants](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_grants/) module:	Manage grants for ClickHouse users and roles
- [clickhouse_info](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_info/) module:	Gather ClickHouse server information using the clickhouse-driver Client interface
- [clickhouse_role](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_role/) module:	Creates or removes a ClickHouse role.
- [clickhouse_user](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/content/module/clickhouse_user/) module:	Creates or removes a ClickHouse user using the clickhouse-driver Client interface

We still need to build a community of users and contributors around the collection. I'd be happy to provide guidance and help.
Please watch, start and look at the issues in our [repo](https://github.com/ansible-collections/community.clickhouse/tree/main).

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.10.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v0-10-0) has been released!

Contains an improvement and a bugfix for the `clickhouse_user` module. See the [changelog](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v0-10-0) for details.

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.11.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v0-11-0) has been released!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.12.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v0-12-0) has been released!

**Comment by Andersson007** [comment]
I think we are ready to start preparations for inclusion of this collection to the Ansible community package (installed via pip install ansible or dnf/apt install ansible). In that case, apart from becoming a part of Ansible's "batteries included", the collection docs will appear on docs.ansible.com.
If you think it's a good idea, please put a like or a comment in https://github.com/ansible-collections/community.clickhouse/issues/137 
Thanks

**Comment by fako1024** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [0.12.1](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v0-12-1) has been released!

**Comment by fako1024** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [1.0.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v1-0-0) has been released!

**Comment by Andersson007** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [1.1.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v1-1-0) has been released!

**Comment by fako1024** [comment]
The [community.clickhouse](https://galaxy.ansible.com/ui/repo/published/community/clickhouse/) collection version [2.0.0](https://github.com/ansible-collections/community.clickhouse/blob/main/CHANGELOG.rst#v2-0-0) has been released!

**Comment by Andersson007** [comment]
the collection is now in the Ansible package https://github.com/ansible-community/ansible-build-data/pull/646#issuecomment-3786052262 , so congrats and thanks to everyone involved!
The collection and its docs will soon appear on docs. ansible.com in the collection index and will be installed automatically with `pip/os_package_manager install ansible`!
